xdsl/transforms/jax_use_donated_arguments.py

```diff
@@ -0,0 +1,143 @@
+import itertools
+from collections.abc import Sequence
+from dataclasses import dataclass
+
+from xdsl.context import MLContext
+from xdsl.dialects import builtin
+from xdsl.dialects.bufferization import MaterializeInDestinationOp
+from xdsl.dialects.builtin import ArrayAttr, DictionaryAttr, FunctionType, TensorType
+from xdsl.dialects.func import FuncOp, ReturnOp
+from xdsl.ir import BlockArgument, Operation, SSAValue
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+def map_donated_input_by_output(
+    donatable_inputs: Sequence[BlockArgument], outputs: Sequence[SSAValue]
+) -> dict[SSAValue, BlockArgument]:
+    """
+    Find suitable donated buffers for each of returned variables.
+    Each buffer can be used only once.
+    Types of the buffer and the variable should match.
+    """
+
+    donatable_inputs_by_type = {
+        type: tuple(group)
+        for type, group in itertools.groupby(
+            sorted(donatable_inputs, key=lambda i: str(i.type)), lambda i: i.type
+        )
+    }
```

Comment:
Yeah I think with the groupby deficiency something like this might be a little easier to read:
```suggestion
    donatable_inputs_by_type = defaultdict(list)
    for value in donatable_inputs:
        donatable_inputs_by_type[value.type].append(value)
    }
```